### PR TITLE
Worldgen Structure Placement

### PR DESCRIPTION
**Pull request by jacobsjo**
An initial implementation of structure placement. Completely untested and is still missing some parts.

What already exists:
- `StructurePlacement` class that does the chunk selection. The `RandomSpreadStructurePlacement` is fully implemented
- `StructureSet` class to handle selection and placement of the structures
- `WorldgenStructure` the structure types themselfes. Needed only for the `findGenerationPoint` method that determines the position within the chunk 
- `HeightProvider` implementation of the Height provider, needed for the jigsaw starting position configuration
- `HolderSet<T>` implementation of Tags. Needed for the biome tag for allowed biomes of the structures.
- Registries now have a `getTagRegistry()` function that returns a new Registry to hold the tags of the base type (`Registry<HolderSet<T>>`)

What is still missing:
- [x] `ConcentricRingsStructurePlacement` placement
- [ ] `findGenerationPoint` for:
    - [x] `JigsawStructure`: including template pools and structure nbt
    - [x] `EndCityStructure`
    - [x] `WoodlandMansionStructure`
    - [ ] ~~`MineshaftStructure`~~: requires complete generation to determine placement
    - [ ] `NetherFossilStructure`: requires access to noise, maybe doable, probably not worth it though
    - [ ] `OceanMonumentStructure`: probably just bad performance from lots of biome access
    - [ ] `RuinedPortalStructure`: very long code, probably not worth it
- [x] implementation of `NoiseBasedChunkGenerator.getBaseHeight`
- [x] Optional entries in `HolderSet` (tags)
- [x] `ExclusionZone`
- [ ] Tests
    - [x] HeightProvider
    - [x] `RandomSpreadStructurePlacement`
    - [x] HolderSet 
    - [x] `FrequencyRecucer`
    - [x] `ExclusionZone`
    - [x] `ConcentricRingsStructurePlacement`
    - [ ] more?

### CODE REVIEW

**Comment by jacobsjo** [comment]
I've implemented the jigsaw starting pool now, including the template nbt reading. I've tested this on my map with vanilla data to confirm that the generation is identical to ingame.

It is hard to write tests for that though, as this requires so much data to work.

File: src/core/Registry.ts [hunk 1]
```diff
@@ -86,4 +91,22 @@ export class Registry<T> {
...
+		}
+		return this.tags
+	}
+
+}
+
+export namespace Registry{
+	export const REGISTRY = new Registry<Registry<unknown>>(Identifier.create('root'))
+
+	export function register<T>(name: string, parser?: (obj: unknown) => T) {
```
Comment (misode) [inline src/core/Registry.ts:107]: What's the reason to change this from static fields to a namespace?

Comment (jacobsjo) [reply src/core/Registry.ts]: changed to static again field again

File: src/math/random/LegacyRandom.ts [hunk 2]
```diff
@@ -71,10 +71,26 @@ export class LegacyRandom implements Random {
 	}
 }
 
+export namespace LegacyRandom {
+	export function fromLargeFeatureSeed(worldSeed: bigint, x: number, z: number): LegacyRandom {
```
[the remark this answers is not in the document]
Comment (jacobsjo) [reply src/math/random/LegacyRandom.ts]: moved to static method.

**Comment by jacobsjo** [comment]
new `ConcentricRingsStructurePlacement` test is failing because of some issue in (probably) in `BiomeSource.findBiomeHorizontal` that leads to it not be in parity with vanilla that I can't figure out yet.

**Comment by jacobsjo** [comment]
Ok, I think this is ready now. The only thing that is theoretically still missing is the `findGenerationPoint` for a few structure types. Those are rather complicated and probably not worth the effort.

**Comment by misode** [comment]
@jacobsjo I've published a beta version of this as `0.18.0-beta.1` because I would like to get a more detailed look.  Sorry for that force push, I accidentally pushed the commit that changed the version, which caused conflicts.

**Review by misode — APPROVED** [review]
I got part of this code working for my use case, and the code looks pretty good 👍 

There are still some messy areas like how some registries are in `WorldgenRegistries` and others are not, but those are things that can be resolved in the future.

**Comment by jacobsjo** [comment]
Thanks! :+1: 

The `WorldgenRegistries` was giving me issues with circular dependencies. I think you had a similar issue with density functions but your work-around only works once. I think it would make more sense to put all the registries into their respective classes.

**Comment by misode** [comment]
> I think it would make more sense to put all the registries into their respective classes.

Yes that seems like a sensible solution.

**Pull request merged**